repose-aggregator/components/filters/keystone-v2-filter/src/main/scala/org/openrepose/filters/keystonev2/KeystoneV2AuthorizationFilter.scala

```diff
@@ -0,0 +1,131 @@
+    // TODO: Stop using request attributes. Stop using the token. Just use the tenant to roles map, tenant ids, and roles headers.
+    getToken(request) flatMap { token =>
+      doAuthorization(configuration, request, token, getEndpoints(request)) match {
+        case AuthorizationPassed(scopedToken, Some(matchedTenant)) =>
+          scopeTenantIdHeader(request, matchedTenant)
+          scopeRolesHeader(request, scopedToken)
+          Success(Unit)
+        case AuthorizationPassed(_, None) => Success(Unit)
+        case AuthorizationFailed(_, _, exception) => Failure(exception)
+      }
+    }
+  }
+
+  def getToken(request: ServletRequest): Try[ValidToken] = {
+    Try {
+      Option(request.getAttribute(TokenRequestAttributeName)).get.asInstanceOf[ValidToken]
+    } recover {
+      case nsee: NoSuchElementException => throw MissingTokenException("Token request attribute does not exist", nsee)
+      case cce: ClassCastException => throw InvalidTokenException("Token request attribute is not a valid token", cce)
+    }
+  }
+
+  def getEndpoints(request: HttpServletRequest): Try[EndpointsData] = {
+    Try {
+      val jsonString = Option(request.getHeader(X_CATALOG))
+        .map(Base64.getDecoder.decode)
+        .map(new String(_))
+        .get
+      val json = Json.parse(jsonString)
+
+      (json \ "endpoints").validate[Vector[Endpoint]]
+        .map(EndpointsData(jsonString, _))
+        .getOrElse(throw new JsonProcessingException("Could not validate endpoints JSON") {})
+    } recover {
+      case nsee: NoSuchElementException =>
+        throw MissingEndpointsException(s"$X_CATALOG header does not exist", nsee)
+      case e@(_: IllegalArgumentException | _: JsonProcessingException) =>
+        throw InvalidEndpointsException(s"$X_CATALOG header value is not a valid endpoints representation", e)
+    }
+  }
+
+  def scopeTenantIdHeader(request: HttpServletRequestWrapper, matchedTenant: String): Unit = {
+    val tenantHandling = Option(configuration.getTenantHandling)
+    val sendAllTenantIds = tenantHandling.exists(_.isSendAllTenantIds)
+    val matchedTenantQuality = tenantHandling.map(_.getSendTenantIdQuality).flatMap(Option.apply).map(_.getUriTenantQuality)
+
+    (sendAllTenantIds, matchedTenantQuality) match {
+      case (true, Some(quality)) => request.addHeader(TENANT_ID, matchedTenant, quality)
+      case (true, None) => request.addHeader(TENANT_ID, matchedTenant)
+      case (false, Some(quality)) => request.replaceHeader(TENANT_ID, matchedTenant, quality)
+      case (false, None) => request.replaceHeader(TENANT_ID, matchedTenant)
+    }
+  }
+
+  def scopeRolesHeader(request: HttpServletRequestWrapper, scopedToken: ValidToken): Unit = {
+    request.removeHeader(ROLES)
+    scopedToken.roles.map(_.name).foreach(request.appendHeader(ROLES, _))
+  }
+}
+
+object KeystoneV2AuthorizationFilter {
+
+  case class MissingTokenException(message: String, cause: Throwable = null) extends Exception(message, cause)
```

Comment:
I'm not a huge fan of all these extending straight exception. We should make a base class that they extend.

Reply:
Fair enough, I'll make the base class.